File: UIKit and Animations/WWDC 2013/Implementing Engaging UI on iOS/Sample_Code/RunningWithASnap/Running with a Snap/LensFlareView.m
/*
     File: LensFlareView.m
 Abstract: 
 
  Version: 1.0
 
 Disclaimer: IMPORTANT:  This Apple software is supplied to you by Apple
 Inc. ("Apple") in consideration of your agreement to the following
 terms, and your use, installation, modification or redistribution of
 this Apple software constitutes acceptance of these terms.  If you do
 not agree with these terms, please do not use, install, modify or
 redistribute this Apple software.
 
 In consideration of your agreement to abide by the following terms, and
 subject to these terms, Apple grants you a personal, non-exclusive
 license, under Apple's copyrights in this original Apple software (the
 "Apple Software"), to use, reproduce, modify and redistribute the Apple
 Software, with or without modifications, in source and/or binary forms;
 provided that if you redistribute the Apple Software in its entirety and
 without modifications, you must retain this notice and the following
 text and disclaimers in all such redistributions of the Apple Software.
 Neither the name, trademarks, service marks or logos of Apple Inc. may
 be used to endorse or promote products derived from the Apple Software
 without specific prior written permission from Apple.  Except as
 expressly stated in this notice, no other rights or licenses, express or
 implied, are granted by Apple herein, including but not limited to any
 patent rights that may be infringed by your derivative works or by other
 works in which the Apple Software may be incorporated.
 
 The Apple Software is provided by Apple on an "AS IS" basis.  APPLE
 MAKES NO WARRANTIES, EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION
 THE IMPLIED WARRANTIES OF NON-INFRINGEMENT, MERCHANTABILITY AND FITNESS
 FOR A PARTICULAR PURPOSE, REGARDING THE APPLE SOFTWARE OR ITS USE AND
 OPERATION ALONE OR IN COMBINATION WITH YOUR PRODUCTS.
 
 IN NO EVENT SHALL APPLE BE LIABLE FOR ANY SPECIAL, INDIRECT, INCIDENTAL
 OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 INTERRUPTION) ARISING IN ANY WAY OUT OF THE USE, REPRODUCTION,
 MODIFICATION AND/OR DISTRIBUTION OF THE APPLE SOFTWARE, HOWEVER CAUSED
 AND WHETHER UNDER THEORY OF CONTRACT, TORT (INCLUDING NEGLIGENCE),
 STRICT LIABILITY OR OTHERWISE, EVEN IF APPLE HAS BEEN ADVISED OF THE
 POSSIBILITY OF SUCH DAMAGE.
 
 Copyright (C) 2013 Apple Inc. All Rights Reserved.
 
 
 Copyright © 2013 Apple Inc. All rights reserved.
 WWDC 2013 License
 
 NOTE: This Apple Software was supplied by Apple as part of a WWDC 2013
 Session. Please refer to the applicable WWDC 2013 Session for further
 information.
 
 IMPORTANT: This Apple software is supplied to you by Apple Inc.
 ("Apple") in consideration of your agreement to the following terms, and
 your use, installation, modification or redistribution of this Apple
 software constitutes acceptance of these terms. If you do not agree with
 these terms, please do not use, install, modify or redistribute this
 Apple software.
 
 In consideration of your agreement to abide by the following terms, and
 subject to these terms, Apple grants you a non-exclusive license, under
 Apple's copyrights in this original Apple software (the "Apple
 Software"), to use, reproduce, modify and redistribute the Apple
 Software, with or without modifications, in source and/or binary forms;
 provided that if you redistribute the Apple Software in its entirety and
 without modifications, you must retain this notice and the following
 text and disclaimers in all such redistributions of the Apple Software.
 Neither the name, trademarks, service marks or logos of Apple Inc. may
 be used to endorse or promote products derived from the Apple Software
 without specific prior written permission from Apple. Except as
 expressly stated in this notice, no other rights or licenses, express or
 implied, are granted by Apple herein, including but not limited to any
 patent rights that may be infringed by your derivative works or by other
 works in which the Apple Software may be incorporated.
 
 The Apple Software is provided by Apple on an "AS IS" basis. APPLE MAKES
 NO WARRANTIES, EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION THE
 IMPLIED WARRANTIES OF NON-INFRINGEMENT, MERCHANTABILITY AND FITNESS FOR
 A PARTICULAR PURPOSE, REGARDING THE APPLE SOFTWARE OR ITS USE AND
 OPERATION ALONE OR IN COMBINATION WITH YOUR PRODUCTS.
 
 IN NO EVENT SHALL APPLE BE LIABLE FOR ANY SPECIAL, INDIRECT, INCIDENTAL
 OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 INTERRUPTION) ARISING IN ANY WAY OUT OF THE USE, REPRODUCTION,
 MODIFICATION AND/OR DISTRIBUTION OF THE APPLE SOFTWARE, HOWEVER CAUSED
 AND WHETHER UNDER THEORY OF CONTRACT, TORT (INCLUDING NEGLIGENCE),
 STRICT LIABILITY OR OTHERWISE, EVEN IF APPLE HAS BEEN ADVISED OF THE
 POSSIBILITY OF SUCH DAMAGE.
 
 EA1002
 5/3/2013
 */

#import "LensFlareView.h"

CGFloat _randomFloat(CGFloat min, CGFloat max) {
    return ((CGFloat)arc4random() / 0x100000000) * (max - min) + min;
}

@interface LensFlareBlob : UIView
- (id)initWithFrame:(CGRect)frame points:(NSInteger)numberOfPoints startAngle:(CGFloat)angle;
@end

@interface LensFlareDiagonalMotionEffect : UIMotionEffect
@property (nonatomic, assign) CGFloat minValue;
@property (nonatomic, assign) CGFloat maxValue;
@end

@interface LensFlareColorMotionEffect : UIMotionEffect 
{
    CGFloat _hue;
    CGFloat _brightness;
}
@property (nonatomic, strong) UIColor *color;
- (instancetype)initWithColor:(UIColor *)color;
@end




@implementation LensFlareView

- (id)initWithFrame:(CGRect)frame {
    return [self initWithFrame:frame flareLineEndPoint:CGPointMake(200, CGRectGetHeight(frame))];
}

- (id)initWithFrame:(CGRect)frame flareLineEndPoint:(CGPoint)endPoint {
    self = [super initWithFrame:frame];
    if (self != nil) {
        [self _addFlareToEndPoint:endPoint];
        self.alpha = _randomFloat(0.15, 0.25);
    }
    return self;
}

// Assumes a right-to-left downward slope starting at {0,0}
- (void)_addFlareToEndPoint:(CGPoint)endPoint {
    CGFloat hypotenuse = sqrtf(powf(endPoint.x,2) + powf(endPoint.y,2));
    CGFloat degrees = atanf(tanf(endPoint.x / endPoint.y)) * 100;
    CGFloat radians = (90.0 - degrees) * M_PI / 180.0;
    
    CGFloat pointOnHypontenuse = _randomFloat(20, 30);
    CGFloat flareSize = 0.0;
    do {
        CGPoint p = CGPointMake(pointOnHypontenuse * cosf(radians), pointOnHypontenuse * sinf(radians));

        // Create a lens flare
        flareSize = _randomFloat(20, 225); 
        LensFlareBlob *blob = [[LensFlareBlob alloc] initWithFrame:CGRectMake(0, 0, flareSize, flareSize) points:7 startAngle:_randomFloat(0, M_PI)];
        blob.center = p;
        
        // Pick a random color and assign it to a new color motion effect
        UIColor *randomColor = [UIColor colorWithHue:_randomFloat(0, 1) saturation:1.0 brightness:_randomFloat(.2, .6) alpha:1.0];
        LensFlareColorMotionEffect *colorMotion = [[LensFlareColorMotionEffect alloc] initWithColor:randomColor];
        
        // Create a new diagonal motion effect
        LensFlareDiagonalMotionEffect *diagonalEffect = [[LensFlareDiagonalMotionEffect alloc] init];
        diagonalEffect.minValue = -20;
        diagonalEffect.maxValue = 30;

        // Group the motion effects together so they get evaluated simultaneously
        UIMotionEffectGroup *group = [[UIMotionEffectGroup alloc] init];
        group.motionEffects = @[colorMotion, diagonalEffect];
        [blob addMotionEffect:group];
        
        [self addSubview:blob];
        
        pointOnHypontenuse += _randomFloat(flareSize * .7, flareSize * .7 + 80);
    } while (pointOnHypontenuse < hypotenuse);
}

@end



@implementation LensFlareBlob

+(Class)layerClass {
    return [CAShapeLayer class];
}

- (id)initWithFrame:(CGRect)frame points:(NSInteger)numberOfPoints startAngle:(CGFloat)angle {
    if (numberOfPoints < 3) {
        NSLog(@"points must be 3 or greater");
        return nil;
    }
    
    // Make us square
    frame.size.height = frame.size.width;
    
    self = [super initWithFrame:frame];
    if (self != nil) {
        self.clipsToBounds = YES;
        NSInteger pointsDrawn = 0;
        UIBezierPath *path = [UIBezierPath bezierPath];
        CGFloat radius = CGRectGetWidth(self.bounds) / 2.0;
        CGPoint p = CGPointMake(radius * cosf(angle) + radius, radius * sinf(angle) + radius);
        [path moveToPoint:p];
        do {
            pointsDrawn++;
            angle += (M_PI * 2) / numberOfPoints;
            p = CGPointMake(radius * cosf(angle) + radius, radius * sinf(angle) + radius);
            [path addLineToPoint:p];
        } while (pointsDrawn < numberOfPoints);
        
        [path closePath];
        
        CAShapeLayer *sl = (CAShapeLayer *)self.layer;
        sl.path = path.CGPath;
    }
    
    return self;
}

@end



@implementation LensFlareDiagonalMotionEffect

- (NSDictionary *)keyPathsAndRelativeValuesForViewerOffset:(UIOffset)viewerOffset {
    // Math!
    CGFloat f = (viewerOffset.horizontal / 2.0) + 0.5;
    CGFloat calculatedValue = self.minValue * (1 - f) + self.maxValue * f;
    
    CGAffineTransform rotationTransform = CGAffineTransformMakeRotation(M_PI_4);
    CGPoint rotatedPoint = CGPointApplyAffineTransform(CGPointMake(calculatedValue, 0), rotationTransform);

    return @{ @"center.x" : @(rotatedPoint.x), @"center.y" : @(rotatedPoint.y) };
}

@end






@implementation LensFlareColorMotionEffect

- (instancetype)initWithColor:(UIColor *)color {
    self = [super init];
    if (self != nil) {
        _color = color;
        
        CGFloat h, b;
        [_color getHue:&h saturation:nil brightness:&b alpha:nil];
        
        _hue = h;
        _brightness = b;
    }
    return self;
}

- (NSDictionary *)keyPathsAndRelativeValuesForViewerOffset:(UIOffset)viewerOffset {
    // Map horizontal movement to brightness and vertical movement to hue. 
    CGFloat hue = 0.0;
    CGFloat brightness = 0.0;
    
    // Math!
    if (viewerOffset.horizontal > 0) {
        brightness = _brightness + (1 - _brightness) * viewerOffset.horizontal;
    }
    else {
        brightness = _brightness + _brightness * viewerOffset.horizontal;
    }
    hue = fabsf(modff(1.0 + _hue + viewerOffset.vertical, &(float){0}));
    
    UIColor *newColor = [UIColor colorWithHue:hue saturation:1.0 brightness:brightness alpha:1.0];
    
    return @{ @"layer.fillColor" : (id)newColor.CGColor };
}

@end

File: UIKit and Animations/WWDC 2013/Implementing Engaging UI on iOS/Sample_Code/Running with a Snap - Swift/Running with a Snap/View Controllers/Launch Screen/LensFlareView.swift
//
//  LensFlareView.swift
//  Running with a Snap
//
//  Created by Seyed Samad Gholamzadeh on 7/11/18.
//  Copyright © 2018 Seyed Samad Gholamzadeh. All rights reserved.
//

import UIKit

func randomFloat(_ min: CGFloat, _ max: CGFloat) -> CGFloat {
//	let random = CGFloat(arc4random_uniform(UInt32(max - min))) + min
	let random = (CGFloat(arc4random()) / 0x100000000) * (max - min) + min
	return random
}

class LensFlareView: UIView {

	override convenience init(frame: CGRect) {
		self.init(frame: frame, flareLineEndPoint: CGPoint(x: 200, y: frame.height))
	}
	
	init(frame: CGRect, flareLineEndPoint endPoint: CGPoint) {
		super.init(frame: frame)
		self.addFlareToEndPoint(endPoint)
		self.alpha = randomFloat(0.15, 0.25)
	}
	
	required init?(coder aDecoder: NSCoder) {
		fatalError("init(coder:) has not been implemented")
	}
	
	// Assumes a right-to-left downward slope starting at {0,0}
	func addFlareToEndPoint(_ endPoint: CGPoint) {
		
		let hypotenuse: CGFloat = sqrt(pow(endPoint.x,2) + pow(endPoint.y,2))
		let degrees: CGFloat = atan(tan(endPoint.x / endPoint.y)) * 100
		let radians: CGFloat = (90.0 - degrees) * CGFloat.pi / 180.0
		
		var pointOnHypontenuse: CGFloat = randomFloat(20, 30)
		var flareSize: CGFloat = 0.0
		
		repeat {
			
			let p: CGPoint = CGPoint(x: pointOnHypontenuse * cos(radians), y: pointOnHypontenuse * sin(radians))
			
			// Create a lens flare
			flareSize = randomFloat(20, 225)
			let blob = LensFlareBlob(frame: CGRect(x: 0, y: 0, width: flareSize, height: flareSize), points: 7, startAngle: randomFloat(0, CGFloat.pi))
			blob?.center = p
			
			// Pick a random color and assign it to a new color motion effect
			let randomColor = UIColor(hue: randomFloat(0, 1), saturation: 1.0, brightness: randomFloat(0.2, 0.6), alpha: 1.0)
			let colorMotion = LensFlareColorMotionEffect(color: randomColor)
			
			// Create a new diagonal motion effect
			let diagonalEffect = LensFlareDiagonalMotionEffect()
			diagonalEffect.minValue = -20
			diagonalEffect.maxValue = 30

			// Group the motion effects together so they get evaluated simultaneously
			let group = UIMotionEffectGroup()
			group.motionEffects = [colorMotion, diagonalEffect]
			blob?.addMotionEffect(group)
			
			self.addSubview(blob!)
			
			pointOnHypontenuse += randomFloat(flareSize * 0.7, flareSize * 0.7 + 80)
		}
		while pointOnHypontenuse < hypotenuse
	}

}

class LensFlareBlob: UIView {
	
	class override var layerClass: AnyClass {
		return CAShapeLayer.self
	}
	
	init?(frame: CGRect, points numberOfPoints: Int, startAngle angle: CGFloat) {
		if numberOfPoints < 3 {
		print("points must be 3 or greater")
			return nil
		}
		
		// Make us square
		var frame = frame
		var angle = angle

		frame.size.height = frame.size.width
		
		super.init(frame: frame)
		
		self.clipsToBounds = true
		var pointsDrawn: Int = 0
		let path = UIBezierPath()
		let radius = self.bounds.width/2.0
		var p = CGPoint(x: radius * cos(angle) + radius, y: radius * sin(angle) + radius)
		path.move(to: p)
		
		repeat {
			pointsDrawn += 1
			angle += (CGFloat.pi * 2) / CGFloat(numberOfPoints)
			p = CGPoint(x: radius * cos(angle) + radius, y: radius * sin(angle) + radius)
			path.addLine(to: p)
			
		} while pointsDrawn < numberOfPoints
		
		path.close()
		
		let sl = (self.layer as! CAShapeLayer)
		sl.path = path.cgPath
		
	}
	
	required init?(coder aDecoder: NSCoder) {
		fatalError("init(coder:) has not been implemented")
	}
}

class LensFlareDiagonalMotionEffect: UIMotionEffect {
	
	var minValue: CGFloat!
	var maxValue: CGFloat!
	
	override func keyPathsAndRelativeValues(forViewerOffset viewerOffset: UIOffset) -> [String : Any]? {
		// Math!
		let f: CGFloat = (viewerOffset.horizontal / 2.0) + 0.5
		let calculatedValue: CGFloat = self.minValue * (1 - f) + self.maxValue * f
		let rotationTransform = CGAffineTransform(rotationAngle: CGFloat.pi/4)
		let rotatedPoint = CGPoint(x: calculatedValue, y: 0).applying(rotationTransform)
		
		return ["center.x" : rotatedPoint.x, "center.y" : rotatedPoint.y]
	}
}

class LensFlareColorMotionEffect: UIMotionEffect {
	
	var hue: CGFloat!
	var brightness: CGFloat!
	var color: UIColor
	
	init(color: UIColor) {
		self.color = color
		super.init()
		
		var h: CGFloat = 0
		var b: CGFloat = 0
		self.color.getHue(&h, saturation: nil, brightness: &b, alpha: nil)
		
		self.hue = h
		self.brightness = b
	}
	
	required init?(coder aDecoder: NSCoder) {
		fatalError("init(coder:) has not been implemented")
	}
	
	override func keyPathsAndRelativeValues(forViewerOffset viewerOffset: UIOffset) -> [String : Any]? {
		// Map horizontal movement to brightness and vertical movement to hue.
		var hue: CGFloat = 0.0
		var brightness: CGFloat = 0.0
		
		// Math!
		if viewerOffset.horizontal > 0 {
			brightness = self.brightness + (1 - self.brightness) * viewerOffset.horizontal
		}
		else {
			brightness = self.brightness + self.brightness * viewerOffset.horizontal
		}
		var floatValue: Float = 0.0
		hue = CGFloat(fabs(modff(Float(1.0 + self.hue + viewerOffset.vertical), &floatValue)))
		let newColor = UIColor(hue: hue, saturation: 1.0, brightness: brightness, alpha: 1.0)
		
		return ["layer.fillColor" : newColor.cgColor]
	}
}



